concrete/src/Entity/User/LoginAttempt.php

```diff
@@ -0,0 +1,94 @@
+<?php
+
+namespace Concrete\Core\Entity\User;
+use Doctrine\ORM\Mapping as ORM;
+
+/**
+ * @ORM\Entity(repositoryClass="Concrete\Core\Entity\User\LoginAttemptRepository")
+ * @ORM\Table(
+ *     name="UserLoginAttempts"
+ * )
+ */
+class LoginAttempt
+{
+
+    /**
+     * The attempt ID, this will be a unique identifier
+     *
+     * @ORM\Id @ORM\Column(type="guid")
+     * @ORM\GeneratedValue(strategy="UUID")
+     */
+    protected $id;
+
+    /**
+     * The user ID that a login was attempted for
+     *
+     * @ORM\Column(type="integer", options={"unsigned": true})
+     */
+    protected $userId;
+
+    /**
+     * The datetime of the time the login happened. This date must be in UTC
+     *
+     * @ORM\Column(type="datetime")
+     */
+    protected $utcDate;
+
+    /**
+     * @return mixed
+     */
+    public function getId()
+    {
+        return $this->id;
+    }
+
+    /**
+     * @param mixed $id
+     *
+     * @return LoginAttempt
```

Comment:
I'd write `@return $this` instead of `@return LoginAttempt` (or `@return self`), because this is not an immutable class

PS: I'm not suggesting that this class should be immutable... I hate immutable classes.

Reply:
I use the class name because it means something different than `self` or `$this`. One is associated with a type, one is specifically saying that it returns the exact instance you're calling on, which isn't exactly what I'm trying to indicate here.

Once we start using return types in v9 our @returns will always be tied to the parent type or an interface so it's probably best to just continue using class names rather than `$this` or `self`.

Reply:
Phpdoc is much more flexible and meaningful than php return types.
By using `@return $this` you explicitly tell developers that this method is not immutable

Reply:
What do you mean by "this method is not immutable"? Do you mean that the return type is effectively mixed?

Reply:
Another example why I wouldn't use `* @return LoginAttempt`. Assume this code:

```php
class LoginAttempt
{
    /** @return LoginAttempt */
    public function setId($id)
    {
        return $this;
    }
}
class MyLoginAttempt extends LoginAttempt 
{
    public function myMethod()
    {
    }
}

$instance = new MyLoginAttempt();
$instance->setId(1)->|
```
IDEs won't be able to suggest `myMethod`, because the IDE sees that `setId` returns a `LoginAttempt` instance.